napari/plugins/theme.py

```diff
@@ -0,0 +1,155 @@
+"""Handling of theme data by the hook specification."""
+from logging import getLogger
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+from napari_plugin_engine import HookImplementation, PluginCallError
+
+from ..utils.translations import trans
+from . import plugin_manager
+
+logger = getLogger(__name__)
+
+
+def get_stylesheet_from_plugins(
```

Comment:
It looks like you loosely patterned `get_stylesheet_from_plugins` after our reader plugin "call hook"... which is probably not the best one to use here.  That one loops through all of the plugins every time, but for this one, it shouldn't be necessary to perform this functionality any more than once per package.  For that, the plugin engine provides the "call_historic" approach, which essentially let's actually "call" the hook a single time (at load) and then when each plugin gets registered for the first time, the hook will be called for that specific plugin.  See, for example how we register the `dock_widgets`, starting [here](https://github.com/napari/napari/blob/ed6b8ef2e55471b42a3ea47875b7e1ef0f99ba80/napari/plugins/_plugin_manager.py#L419)

1. `discover_widgets` calls the actual hook with `self.hook.<specname>.call_historic()`.  this is in contrast to iterating through `get_hookimpls()` as we do with reader plugins.  The _default_ behavior of the plugin engine is actually to call all the hooks, so that whole looping we do in the reader plugin is a special/odd case.  (note, the call to `discover_widgets` actually happens only once our `qt_main_window` is instantiated for the first time).
2. Each time a plugin is registered, if it provides a `<specname>` implementation, the function provided to `call_historic` will be called (in the case of dock widgets, it's [`register_dock_widget`](https://github.com/napari/napari/blob/ed6b8ef2e55471b42a3ea47875b7e1ef0f99ba80/napari/plugins/_plugin_manager.py#L304)).  The correlate here would be a function that is responsible for performing the qss & theme dict aggregation, and compiling the resources.  (It would also do all the return value validation).

This _would_ essentially mean that a lot of the resource compilation would need to be refactored a little bit.  That is, rather than gathering _all_ of the icons each time and recompiling them anytime anything is different, the callback provided to `call_historic` would ideally be able to compile just the new svgs for each plugin.  (I'm not sure if that would come with much of a performance penalty... but it would make it cleaner to register each plugin individually).

Reply:
Thanks for the pointers and ideas. I think with small number of theme/stylesheet the current approach would lead to significant performance loss but it can be cleaner. I will look into changing the behaviour to discover themes etc once at the startup rather than recompile all of the resources each time a new plugin is installed.